kratos/utilities/parallel_utilities.h

```diff
@@ -143,10 +143,10 @@ class BlockPartition
      * the function f needs to return the values to be used by the reducer
      * @param TReducer template parameter specifying the reduction operation to be done
      * @param TThreadLocalStorage template parameter specifying the thread local storage
-     * @param f - must be a unary function accepting as input TContainerType::value_type& and the thread local storage
+     * @param f - must be a function accepting as input TContainerType::value_type& and the thread local storage
      */
-    template <class TReducer, class TThreadLocalStorage, class TUnaryFunction>
-    inline typename TReducer::value_type for_each(const TThreadLocalStorage rThreadLocalStoragePrototype, TUnaryFunction &&f)
+    template <class TReducer, class TFunction, class TThreadLocalStorage>
+    inline typename TReducer::value_type for_each(TFunction &&f, const TThreadLocalStorage rThreadLocalStoragePrototype)
```

Comment:
I would argue that you can use & in the parameter here in order to avoid 1 copy

Reply:
in my original proposal I intentionally passed by value to have the "private" working.
i.e. if I pass by ref I cannot construct in place:
~~~cpp
BlockPartition<std::vector<double>>(data_vector).for_each(ClassBasedTLS(),...);
~~~